evm/src/executor/inspector/tracer.rs

```diff
@@ -286,7 +287,7 @@ where
         self.fill_trace(
             status,
             gas_used(data.env.cfg.spec_id, gas.spend(), gas.refunded() as u64),
-            code,
+            if success { code } else { retdata.to_vec() },
```

Comment:
this is independent of the status because in the trace object, `result` and `error` are mutually exclusive, failed tx should have `error` and succeeding txs have `result`

so we actually need to check the return status here:

https://github.com/foundry-rs/foundry/blob/91f69dd546e45450c393a3edfa0e215c93458706/anvil/src/eth/backend/mem/storage.rs#L383-L391

rn we always set result

this means we need to make a few changes to the `CallTrace` object I suppose, so it can distinguish between `code` and `output`, 
perhaps just `Option<Bytes>` for code? which is then used in `parity_result()`?

Reply:
I don't see a reason for adding a new field to the `CallTrace` struct.

In this function, `retdata` contain revert data or return data (mutually exclusive), depending on the status:
https://github.com/foundry-rs/foundry/blob/91f69dd546e45450c393a3edfa0e215c93458706/evm/src/executor/inspector/tracer.rs#L228-L245

Even from the EVM perspective, contract code and revert data are stored at the same data location (return data) - see https://www.evm.codes/#f0?fork=merge and https://www.evm.codes/#f0?fork=merge. Return data contain either the contract code or revert reason data, depending on the status.

I believe the `error` field in `Trace` serves a different purpose. It should contain human-readable info about why the trace failed. One of these reasons should be "Reverted". The `result` field should contain the return value - the data contained in `return data` data storage (see https://www.evm.codes/about#returndata).
This can be return value or revert data (mutually exclusive) for non-CREATE traces and contract code or revert data (also mutually exclusive) for CREATE traces.

TL;DR
I would leave the `CallTrace` struct with just one field (named `output`) because this is how EVM works (`return data` data storage). The `output` field would be used accordingly in the [parity_result](https://github.com/foundry-rs/foundry/blob/91f69dd546e45450c393a3edfa0e215c93458706/evm/src/trace/node.rs#L49-L63) function depending on the trace `status`. But there is currently no way to set the revert data for CREATE traces and the `error` field in `Trace` serves a different purpose.

Perhaps to make it clearer, I can also make a PR to the [ethers-rs](https://github.com/gakonst/ethers-rs) repository that will be needed prior to this PR anyway.

Reply:
The corresponding pull request in ethers-rs: https://github.com/gakonst/ethers-rs/pull/2392